### PR TITLE
[Merged by Bors] - feat: new file `Data/List/Pi`

### PR DESCRIPTION
**Pull request by astrainfinita**
The new APIs replace some definitions and lemmas in `Data.FinEnum`. This PR also provides more lemmas.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/List/Pi.lean [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+lemma cons_def (a : α i) (f : ∀ j ∈ l, α j) : cons a f =
+    fun j hj ↦ if h : j = i then h.symm.rec a else f j <| (mem_cons.1 hj).resolve_left h :=
+  rfl
+
+@[simp] lemma _root_.Multiset.Pi.cons_coe {l : List ι} (a : α i) (f : ∀ j ∈ l, α j) :
+    Multiset.Pi.cons (ι := ι) (m := l) a f = cons a f :=
+  rfl
+
+@[simp] lemma cons_eta (f : ∀ j ∈ (i :: l), α j) :
```
Comment (YaelDillies) [inline Mathlib/Data/List/Pi.lean:46]:
```suggestion
@[simp] lemma cons_eta (f : ∀ j ∈ i :: l, α j) :
```

File: Mathlib/Data/Multiset/Pi.lean [hunk 2]
```diff
@@ -16,74 +16,97 @@ namespace Multiset
...
 
-/-- Given `δ : α → Type*`, a multiset `m` and a term `a`, as well as a term `b : δ a` and a
-function `f` such that `f a' : δ a'` for all `a'` in `m`, `Pi.cons m a b f` is a function `g` such
-that `g a'' : δ a''` for all `a''` in `a ::ₘ m`. -/
-def Pi.cons (m : Multiset α) (a : α) (b : δ a) (f : ∀ a ∈ m, δ a) : ∀ a' ∈ a ::ₘ m, δ a' :=
-  fun a' ha' => if h : a' = a then Eq.ndrec b h.symm else f a' <| (mem_cons.1 ha').resolve_left h
+/-- Given `α : ι → Sort*`, a multiset `m` and a term `i`, as well as a term `a : α i` and a
+function `f` such that `f j : α j` for all `j` in `m`, `Pi.cons a f` is a function `g` such
+that `g k : α k` for all `k` in `i ::ₘ m`. -/
+def cons (a : α i) (f : ∀ j ∈ m, α j) : ∀ j ∈ (i ::ₘ m), α j :=
```
Comment (YaelDillies) [inline Mathlib/Data/Multiset/Pi.lean:35]: Are you sure it's a good idea to make `i` and `m` implicit here? What's your reasoning?

**Comment by digama0** [comment]
Note: I have pushed an update to the lean toolchain because this PR was on a [buggy version](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Potential.20data.20loss.20from.20.60lake.20clean.60.20with.204.2E2.2E0-rc2.2F3/near/397875701) of the toolchain. **WARNING**: checking out old commits of this PR using v4.2.0-rc2 or v4.2.0-rc3 can cause `lake clean` to delete your mathlib folder! If you need to do so, make sure to delete `lakefile.olean` manually before running any `lake` commands.

File: Mathlib/Data/List/Pi.lean [hunk 3]
```diff
@@ -0,0 +1,98 @@
+/-
+Copyright (c) 2023 Yuyang Zhao. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yuyang Zhao
+-/
+import Mathlib.Data.Multiset.Pi
+
+/-!
+# The cartesian product of lists
```
Comment (urkud) [inline Mathlib/Data/List/Pi.lean:9]: Please write the module docstring.

**Comment by tobiasgrosser** [comment]
@negiizhao, I am curious what the status of this PR is. Do you plan to merge this PR still?

**Comment by negiizhao** [comment]
I'll come back to this PR when I have time. `Data.Multiset.Pi` and `Data.Finset.Pi` also don't have the module docstring so I can't fix it quickly.

**Comment by tobiasgrosser** [comment]
OK. We had code that uses this, so merging this in would actually be pretty convenient for us.

**Comment by YaelDillies** [comment]
@tobiasgrosser and co, maybe you can take over this PR if @negiizhao agrees?

**Review by YaelDillies — COMMENTED** [review]
I'm really surprised `Data.List.Pi` imports `Data.Multiset.Pi` and not the other way around. Can you explain?

**Review by YaelDillies — COMMENTED** [review]
If Eric asked for it, I will trust him. The rest looks agreeable.

maintainer merge

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**